ide/rust/SyncRunner.kt

```diff
@@ -0,0 +1,128 @@
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.BAZEL_BIN
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.BUILD
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.INFO
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.OUTPUT_BASE
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.OUTPUT_GROUPS_RUST_IDE_SYNC
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.QUERY
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.RUST_IDE_SYNC_ASPECT
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.RUST_TARGETS_DEPS_QUERY
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.RUST_TARGETS_QUERY
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.VATICLE_REPO_PREFIX
+import picocli.CommandLine
+import java.io.File
+import java.nio.file.Path
+import java.util.concurrent.Callable
+import kotlin.io.path.Path
+import kotlin.system.exitProcess
+
+fun main(args: Array<String>): Unit = exitProcess(CommandLine(SyncRunner()).execute(*args))
+
+@CommandLine.Command(name = "sync", mixinStandardHelpOptions = true)
+class SyncRunner : Callable<Unit> {
+
+    @CommandLine.Option(names = ["--verbose", "-v"], required = false)
+    private var verbose: Boolean = false
+
+    private lateinit var logger: Logger
+    private lateinit var shell: Shell
+    private val buildWorkspaceDir = Path(System.getenv("BUILD_WORKSPACE_DIRECTORY"))
+    private lateinit var bazelOutputBase: Path
+
+    override fun call() {
+        logger = Logger(logLevel = if (verbose) DEBUG else ERROR)
+        shell = Shell(logger, verbose)
+        bazelOutputBase = Path(shell.execute(listOf(BAZEL, INFO, OUTPUT_BASE), buildWorkspaceDir).outputString().trim())
+        val rustTargets = rustTargets(buildWorkspaceDir)
+        validateBuildWorkspace(rustTargets)
+        loadRustToolchainAndExternalDeps(rustTargets)
+        vaticleRustRepoPaths().forEach { sync(it) }
+    }
+
+    private fun validateBuildWorkspace(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets + "--build=false", baseDir = buildWorkspaceDir)
+    }
+
+    private fun loadRustToolchainAndExternalDeps(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets, baseDir = buildWorkspaceDir, throwOnError = false)
+    }
+
+    private fun vaticleRustRepoPaths(): List<Path> {
+        // e.g: [/Users/root/workspace/typedb-client-rust, /private/var/_bazel_root_/123abc/external/vaticle_typedb_protocol]
+        return listOf(buildWorkspaceDir) + shell.execute(listOf(BAZEL, QUERY, RUST_TARGETS_DEPS_QUERY, "--output=package"), buildWorkspaceDir)
+            .outputString().split("\n")
+            .filter { it.startsWith(VATICLE_REPO_PREFIX) }
+            .map { it.split("@")[1].split("//")[0] }
+            .map { bazelOutputBase.resolve("external").resolve(it) }
+    }
+
+    private fun sync(repoPath: Path) {
```

Comment:
Similar to how we should rename `vaticleRustRepoPaths`, I think we can rename `repoPath` to just `repo` or `repository.`

Reason: It seems that repositories are always represented as path objects anyway, so referring to them as "repo path" feels redundant.